loadbalancer/algorithm.go

```diff
@@ -89,19 +95,71 @@ func (*consistentHash) Apply(ctx *routing.LBContext) routing.LBEndpoint {
 	key := net.RemoteHost(ctx.Request).String()
 	if _, err := h.Write([]byte(key)); err != nil {
 		log.Errorf("Failed to write '%s' into hash: %v", key, err)
-		return ctx.Route.LBEndpoints[rand.Intn(len(ctx.Route.LBEndpoints))]
+		return ctx.Route.LBEndpoints.At(rand.Intn(ctx.Route.LBEndpoints.Length()))
 	}
 	sum = h.Sum32()
-	choice := int(sum) % len(ctx.Route.LBEndpoints)
+	choice := int(sum) % ctx.Route.LBEndpoints.Length()
 	if choice < 0 {
-		choice = len(ctx.Route.LBEndpoints) + choice
+		choice = ctx.Route.LBEndpoints.Length() + choice
+	}
+	return ctx.Route.LBEndpoints.At(choice)
+}
+
+type powerOfChoices struct {
+	rand            *rand.Rand
+	numberOfChoices int
+	mutex           sync.Mutex
+}
+
+func (a *powerOfChoices) GetScore(e routing.LBEndpoint) int {
+	// TODO: support more metrics. E.g., least connections.
+	// endpoints with higher inflight request should have lower score
+	return -e.Metrics.GetInflightRequests()
+}
+
+// newPowerOfChoices selects N random backends and picks the one with less outstanding requests.
+func newPowerOfChoices(endpoints []string) routing.LBAlgorithm {
+	t := time.Now().UnixNano()
+	return &powerOfChoices{
+		rand:            rand.New(rand.NewSource(t)),
+		numberOfChoices: 2, // TODO: make this the value part of skipper configuration.
+	}
+}
+
+func (a *powerOfChoices) GetRandomIndex(length int) int {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	return a.rand.Intn(length)
+}
+
+func (a *powerOfChoices) Apply(ctx *routing.LBContext) routing.LBEndpoint {
+	numEndpoints := ctx.Route.LBEndpoints.Length()
+
+	// no need to compute a random endpoint if only one is given
+	if numEndpoints == 1 && a.numberOfChoices == 1 {
```

Comment:
`a.numberOfChoices == 1` is redundant - there is no need to draw random if there is only one endpoint

Reply:
I am not sure how the routes are created in skipper ingress. If could happen a created route would have a single endpoint. Also, what happens if an endpoint is marked as sick and the route starts to have a single endpoint.
In any case, this is a micro-op. I removed that for the sake of readability.

Reply:
I mean `if numEndpoints == 1` is fine but `&& a.numberOfChoices == 1` is redundant - it does not make sense to draw random N times if you can pick only from one endpoint.
 
> In any case, this is a micro-op. I removed that for the sake of readability.

This is the best outcome in my opinion :)